src/AssetsListener.php

```diff
@@ -0,0 +1,367 @@
+            throw new Exception\RuntimeException('can not create folder for caching asset');
+        }
+
+        if (is_string($source)) {
+            if (@file_put_contents($targetFile, $source) !== false) {
+                return $source;
+            }
+            throw new Exception\RuntimeException('can not save file to cache');
+        }
+
+        if (is_resource($source)) {
+            $target = fopen($targetFile, 'x+');
+            if (@stream_copy_to_stream($source, $target)) {
+                fclose($source);
+                fseek($target, 0);
+                return $target;
+            }
+            throw new Exception\RuntimeException('can not save file to cache');
+        }
+
+        throw new Exception\InvalidArgumentException(sprintf(
+            '%s: expects "$source" parameter an string or resource, received "%s"',
+            __METHOD__,
+            (is_object($source) ? get_class($source) : gettype($source))
+        ));
+    }
+
+    public function setCacheToPublic($flag)
+    {
+        $this->isCacheToPublic = (bool)$flag;
+        return $this;
+    }
+
+    public function isCacheToPublic()
+    {
+        return $this->isCacheToPublic;
+    }
+
+    /**
+     * @param string $routeName
+     * @return self
+     */
+    public function setRouteName($routeName)
+    {
+        if (!is_string($routeName)) {
+            throw new Exception\InvalidArgumentException(sprintf(
+                '%s: expects parameter an %s, received "%s"',
+                __METHOD__,
+                'string',
+                (is_object($routeName) ? get_class($routeName) : gettype($routeName))
+            ));
+        }
+        $this->routeName = $routeName;
+        return $this;
+    }
+
+    /**
+     * @param array|RouteStackInterface $router
+     * @return self
+     */
+    public function setRouter($router)
+    {
+        if (!(is_array($router) || $router instanceof RouteInterface)) {
+            throw new Exception\InvalidArgumentException(sprintf(
+                '%s: expects parameter an %s or %s, received "%s"',
+                __METHOD__,
+                'array',
+                RouteInterface::class,
+                (is_object($router) ? get_class($router) : gettype($router))
+            ));
+        }
+        $this->router = $router;
+        return $this;
+    }
+
+    public function setRouterCacheFolder($routerCacheFolder)
+    {
+        $this->routerCacheFolder = trim($routerCacheFolder, '/');
+        return $this;
+    }
+
+    public function setUseInternalRouter($useInternalRouter)
+    {
+        $this->useInternalRouter = (bool)$useInternalRouter;
+        return $this;
+    }
+
+    /**
+     * @return AssetsManager
+     */
+    public function getAssetsManager()
+    {
+        return $this->assetsManager;
+    }
+
+    /**
+     * @param AssetsManager $assetsManager
+     * @return self
+     */
+    public function setAssetsManager(AssetsManager $assetsManager)
```

Comment:
If we move `AssetsManager` injection to the constructor, I doubt we need this after?